File: divide-and-conquer/105m/OptimalSolution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
//执行用时：1 ms, 在所有 Java 提交中击败了99.31% 的用户
//内存消耗：38.4 MB, 在所有 Java 提交中击败了61.18% 的用户
class Solution {
    private HashMap<Integer, Integer> hMap = new HashMap<>();
    private int n;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        n = preorder.length;
        for(int i = 0; i < n; i++) hMap.put(inorder[i], i);
        return merge(preorder, 0, n, 0);
    }
    private TreeNode merge(int[] preorder, int l, int r, int l2){
        if(l >= r) return null;
        TreeNode root = new TreeNode(preorder[l]);
        int rootPos = hMap.get(root.val);
        int mid = l + rootPos - l2 + 1;
        root.left = merge(preorder, l+1, mid, l2);
        root.right = merge(preorder, mid, r, rootPos+1);
        return root;
    }
}

File: graph/236/optimal_solution.cpp
//递归求解，containsOneNode函数求root中是否包含p，q中至少一个节点
//最近公共祖先parent一定满足：要么1.parent=p或q且有一颗子树包含p或q，要么2.parent的左右两颗子树各包含一个p，q中节点
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    TreeNode* ret;
    bool containsOneNode(TreeNode* root, TreeNode* p, TreeNode* q){
        if(!root) return false;
        bool l = containsOneNode(root->left, p, q);
        bool r = containsOneNode(root->right, p, q);
        if((l && r) || ((root->val == p->val || root->val == q->val) && (l || r))) ret = root;
        return l || r || root->val == p->val || root->val == q->val;
    }
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        containsOneNode(root, p, q);
        return ret;
    }
};
